### ISSUE TITLE
Custom Livereload Implementation

### ISSUE THREAD

**Issue by mohislm** [broccolijs/broccoli#256]
Since the live-reload functionality is removed from the recent version, in the Brocfile.js I was trying to use tiny-lr and inject livereload code. That works, however how would I make the change event fire from Brocfile when a file changes?

**Comment by joliss** [comment]
Hm, good question. I don't really see a clean way to do this unless you are using Broccoli programmatically, rather than via the Broccoli core CLI.

For an example of how to do it programmatically, Ember CLI [hooks into the watcher events](https://github.com/ember-cli/ember-cli/blob/26204c790a1093f8f413c48817d7da62ea3ea4dc/lib/tasks/server/livereload-server.js#L42-L44).

**Comment by mohislm** [comment]
That just makes life more complicated. Instead of completely removing live-reload from broccoli, you could have made it activated/deactivated via a config option so that other CLI (i.e. ember-cli) could ignore that and use their own. Since I upgraded broccoli for one of my project, it stopped livereloading, and I was thinking of going back to Gulp or probably bare npm build tool. Were you sure of such drastic change before removing LR?

**Comment by joliss** [comment]
I think LR is an awesome feature that most people will generally want for their development workflow. Removing it from Broccoli core was simply motivated by the fact that LR has a fair number of moving parts, and I didn't think that Broccoli core was the best place to put this kind of complexity.

So while I'm pretty sure that I don't want LR support in Broccoli core, it would be great to see other people make libraries that use Broccoli and also provide LR functionality, just like Ember CLI. E.g. one could even make a "broccoli-with-livereload" package (though this seems a bit cumbersome - there might a better way).

**Comment by a-laughlin** [comment]
Spent today learning broccoli after installing it on a personal project last week.  Feeling a bit frustrated trying to get it working since there isn't a straightforward LR solution yet.

What makes it more confusing is that it's unclear how broccoli serve rebuilds are supposed to work.  There is no documentation for broccoli serve.  Looking at the [code](https://github.com/broccolijs/broccoli/blob/master/lib/server.js), it's unclear where it's getting the directory to serve given ./dev and ./dist subdirectories, both with an index.html in them.  Running a simple `live-server` in the ./dist directory doesn't work since the files don't change for some reason with live-server in one terminal, and broccoli serve in another, even though broccoli serve shows a reload when ./app files change.

LR may be more complex in the code, but integrated watching and LR is a big part of what made broccoli feel simple to use.  Given how common a use-case LR is, an easy solution will likely go a long way toward lowering Broccoli's adoption barriers.  Heading back to grunt for now.

**Comment by mikehaas763** [comment]
@joliss Is an example of a library that uses both something like a gulp set up such as a gulp task that sets up LR and runs Broccoli build? Just trying to guage where you're going with it.

Overall I agree that LR really doesn't belong in Broccoli core. What if someone would rather use BrowserSync (myself) or any other implementations? Unfortunately I don't know enough about broccoli yet to provide any alternative solutions.

**Comment by rtablada** [comment]
@joliss would it be possible to make something like a `broccoli-lr` command that just wraps the broccoli watcher in a live reload server to refresh on watched changes?

I'm not to familiar with using broccoli programatically, but it hopefully doesn't seem like it would be super hard?

**Comment by scottsword** [comment]
I'm with @a-laughlin on a lot of points addressed. Kicking the can down the road to Ember-CLI isn't an option when you aren't in an Ember app. I think Broccoli is great and love the idea of it, but I don't understand why LR would be removed. If thats going to be removed then `broccoli serve` might as well be removed too. Consider my scenario. I have a large grunt file that does many things besides my build. I come across the Broccoli project which touts efficient builds with an integrated serve mechanism. Ah ha! I can now cut the fat from grunt and have it truly just execute tasks for me (test, deploys, etc).  I wrote my first Brocfile and everything went great from the terminal, but there was no LR support. After some digging I found that LR was pulled from the source and there isn't even any documentation for the serve API.

IMO if the expectation of Broccoli is to be lower-level tool and not own LR then that should be more transparent. If I can't use `broccoli serve` for development why even offer the command?

--or--

Add a live reload component. I would be hugely excited, because then broccoli, out of the box could take over my entire build process, without any extraneous wiring done to get my browser in sync with the Broccoli build.

**Comment by stefanpenner** [comment]
@scottsword many non ember users, are just using ember-cli (although it is named strange) it really doesn't couple to ember much at all.

**Comment by rtablada** [comment]
@scottsword if you want just LR implementation I would recommend https://github.com/rtablada/broccoli-lr

**Comment by joliss** [comment]
Yes, I think we all agree we'd like to have a better story for "just broccoli", including LR support and such. Maybe a package separate from broccoli core. Of course it'd be great to share this code with ember CLI. 

The LR that we had in Broccoli core was rather lacking so I don't want to just resurrect it as is, and I don't have time to redo it properly right now.  But this is something I'd like to see in the future.

**Comment by mikehaas763** [comment]
I think it would be great if it was an extensibility point so other implementations could be created as well. Such as one for using browsersync.

**Comment by stefanpenner** [comment]
I've been thinking about this. I believe we should:

* provide something
* unlike ember-cli, that something should simply reuse the existing HTTP (Rather then two ports)
* we should likely improve ember-cli's tree-sync approach of efficiently copying to `dist` and using that changeset as what we pass to the LR client. So it can intelligently reload (hot reload images/fonts styles etc)
